File: Data Structures & Algorithms/buy-and-sell-crypto/submission-8.js
class Solution {
    /**
     * @param {number[]} prices
     * @return {number}
     */
    maxProfit(prices) {
        let max = 0;
        let l = 0;
        for (let r = 1; r < prices.length; r ++) {
            let cur = prices[r]
            if (cur <= prices[l]) {
                l = r 
                continue
            } else {
                max = Math.max((cur - prices[l]), max)
            }
        }

        return max 
    }
}

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-8.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        
        def dfs(i, buying): 
            if i >= len(prices): 
                return 0 

            # skip, do nothing
            res = dfs(i + 1, buying) 

            # if last is buying, then in this dfs, we can sell it
            if buying: 
                res = max(res, dfs(i + 1, False) - prices[i]) 
            else:  
                # buy
                res = max(res, dfs(i + 2, True) + prices[i]) 

            return res 

        return dfs(0, True)